### ISSUE TITLE
multi-node distributed training rank0 hang at dataloader after a few epochs

### ISSUE THREAD

**Issue by LiweiPeng** [pytorch/pytorch#87995]
### 🐛 Describe the bug

I am using pytorch to do multi-GPU distributed training. When I use 1-node 8-GPUs, the training works well. However, when I used 4-nodes 32-GPUs, after N epochs (N can be 1 to 3), the training will freeze. From the debugger, it shows that rank0 is stuck at dataloader place with the following stack. And all other 31 ranks are waiting for rank0.  For this hang case, dataloader uses num_workers=8. When I changed num_workers=4, the job can run longer before hang. When I changed num_workers=0, the job won't hang.

This is 100% reproducible. I ran the same tests about 20 times.

Each time it will stuck at the same code place: rank0 will hang at dataloader, all other ranks will wait because of all_reduce(). When this hangs, rank0 already finished several steps (e.g. X is 5).
It always stuck at rank0.

The stack for rank0:
```
Thread 0x7FA6E7C01740 (idle): "MainThread"
    wait (threading.py:302)
    get (queue.py:170)
    __next__ (fairseq/data/iterators.py:648)
    __iter__ (fairseq/data/iterators.py:59)
    _chunk_iterator (fairseq/data/iterators.py:528)
    __iter__ (fairseq/data/iterators.py:59)
    __iter__ (fairseq/logging/progress_bar.py:256)
    train (fairseq_cli/train.py:274)
    inner (contextlib.py:75)
    main (fairseq_cli/train.py:165)
    distributed_main (fairseq/distributed/utils.py:326)
    call_main (fairseq/distributed/utils.py:352)
    cli_main (fairseq_cli/train.py:499)
    <module> (train.py:14)
```
The stack for other ranks:
```
Thread 0x7F6F32481740 (active): "MainThread"
    _all_reduce_dict (fairseq/distributed/utils.py:661)
    all_reduce_dict (fairseq/distributed/utils.py:667)
    _fast_stat_sync_sum (fairseq/trainer.py:1195)
    _aggregate_logging_outputs (fairseq/trainer.py:1135)
    train_step (fairseq/trainer.py:708)
    inner (contextlib.py:75)
    train (fairseq_cli/train.py:278)
    inner (contextlib.py:75)
    main (fairseq_cli/train.py:165)
    distributed_main (fairseq/distributed/utils.py:326)
    call_main (fairseq/distributed/utils.py:352)
    cli_main (fairseq_cli/train.py:499)
    <module> (train.py:14)
```

### Versions

pytorch 1.11
OS: Linux Ubuntu 20.4
Python version: 3.8
CUDA: 11.5
GPU: A100 80GB
fairseq: 1.0.0a0

**Comment by ejguan** [comment]
Could you provide minimum reproducible code for us? Since Rank0 has already advanced for a several steps, I feel like this is related to unbalanced data sharding, which causes hanging at gradient aggregation.

**Comment by LiweiPeng** [comment]
Thanks @ejguan for the quick response. I couldn't provide reproducible code without disclosing all the model internals, which we are not ready yet. 
Using the debugger, I found when the hang occurs, rank0 is at step N (e.g. N could be 5 but not always), and all other ranks are at step N+1. So all other ranks are waiting for rank0 to finish. rank0 is stuck at the dataloader step, as shown above.

**Comment by ejguan** [comment]
DataLoader step is a little bit vague to me. Could you please share where is the hanging happens within dataloader?
The only distributed step within `DataLoader` is here
https://github.com/pytorch/pytorch/blob/1e5d33b6dfc0ed477fec57aca63427d038751207/torch/utils/data/dataloader.py#L570. And, it would only be invoked at the beginning of each epoch.

Unless you have distributed call within your Dataset code, I don't know why this hanging happens within DataLoader.

**Comment by NivekT** [comment]
@ejguan Also worth mentioning that `_share_dist_seed` is only called if `IterDataPipe` is being used.

**Comment by LiweiPeng** [comment]
@ejguan Thanks for pointing me to the code. I was not clear in the previous post. What I found out is: 1) when dataloader uses num_workers=0, there is no hang; 2) the more num_workers I used, the easier and faster it would hang. e.g. when num_workers=8, it takes 7hours. when num_workers=4, it takes 30 hours. 

When it hangs, the code stack didn't show dataloader code. It shows stack like below where rank0 stuck at the data iterator():
```
Thread 0x7FA6E7C01740 (idle): "MainThread"
    wait (threading.py:302)
    get (queue.py:170)
    __next__ (fairseq/data/iterators.py:648)
    __iter__ (fairseq/data/iterators.py:59)
    _chunk_iterator (fairseq/data/iterators.py:528)
    __iter__ (fairseq/data/iterators.py:59)
    __iter__ (fairseq/logging/progress_bar.py:256)
    train (fairseq_cli/train.py:274)
    inner (contextlib.py:75)
    main (fairseq_cli/train.py:165)
    distributed_main (fairseq/distributed/utils.py:326)
    call_main (fairseq/distributed/utils.py:352)
    cli_main (fairseq_cli/train.py:499)
    <module> (train.py:14)
```

**Comment by LiweiPeng** [comment]
This only occurs at the beginning first few steps of certain epoch (not every epoch). I'll double-check how the sharding part of the code works.
